python3 build

I'm trying to build DPPP in a native Python 3 environment, i.e. python == python3.7 and the appropriate boost libraries are all there. (In detail this is a singularity image based on Debian buster.)

With the default makefile I get linking errors of the form
```
100%] Linking CXX executable DPPP
[100%] Built target dppp_testdyndppp
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_Size'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_FromStringAndSize'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_FromString'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyInt_Type'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyInt_FromLong'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyFile_FromString'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyNumber_Divide'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyUnicodeUCS4_FromEncodedObject'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyNumber_InPlaceDivide'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_InternFromString'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyUnicodeUCS4_AsWideChar'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyInt_AsLong'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_AsString'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_FromFormat'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyClass_Type'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyFile_AsFile'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `Py_InitModule4_64'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libboost_python.so: undefined reference to `PyString_Type'
collect2: error: ld returned 1 exit status
make[2]: *** [DPPP/CMakeFiles/DPPP.dir/build.make:413: DPPP/DPPP] Error 1
make[1]: *** [CMakeFiles/Makefile2:514: DPPP/CMakeFiles/DPPP.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

This I think is a python 2/3 incompatibility. I can fix the compile errors by editing CMakeLists.txt to look for Python and explicitly use the version of python it finds for the boost library:

```
FIND_PACKAGE(Python REQUIRED COMPONENTS Interpreter Development)
INCLUDE_DIRECTORIES( ${Python_INCLUDE_DIRS} )

#Prevent accidentally finding old BoostConfig.cmake file from casapy
set(Boost_NO_BOOST_CMAKE ON)
find_package(Boost COMPONENTS date_time filesystem python${Python_VERSION_MAJOR}${Python_VERSION_MINOR} program_options system unit_test_framework REQUIRED)
include_directories(${Boost_INCLUDE_DIR})
```
but this gives a warning about not being able to build the Python DPPP interface (which is fine for my purposes). I don't understand cmake well enough to debug this further.

It would be good if DPPP could cope with a python3 based system...

--- a/.travis/Dockerfile_env
+++ b/.travis/Dockerfile_env
@@ -11,7 +11,7 @@ RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
     gfortran flex bison wcslib-dev \
     libboost-numpy-dev liblua5.3-dev
 
-# Build casacore from source
+# Get casacore
 RUN mkdir /casacore && cd /casacore && git clone -b v3.2.0 https://github.com/casacore/casacore.git src
 RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j24
 
@@ -21,9 +21,3 @@ RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFI
 
 ADD . /src
 WORKDIR /src
-
-RUN mkdir /build && cd /build && cmake ../src
-RUN cd /build && make -j24
-RUN cd /build && make install -j24
-RUN DPPP
-RUN cd /build && CTEST_OUTPUT_ON_FAILURE=1 make test -j24

--- a/.travis/Dockerfile_lofar
+++ b/.travis/Dockerfile_lofar
@@ -1,20 +1,34 @@
-FROM kernsuite/base:4
+FROM ubuntu
 
 RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
-    apt-get install -y casacore-data casacore-dev \
+    apt-get install -y \
     libboost-python-dev libcfitsio-dev python3-dev python3-numpy cmake \
-    build-essential libhdf5-serial-dev libarmadillo-dev lofar-dev \
+    build-essential libhdf5-serial-dev libarmadillo-dev \
     libboost-filesystem-dev libboost-system-dev libboost-date-time-dev \
     libboost-signals-dev libboost-program-options-dev libboost-test-dev \
-    libxml2-dev libpng-dev pkg-config aoflagger-dev \
-    libgtkmm-3.0-dev git wget libfftw3-dev
+    libxml2-dev libpng-dev pkg-config \
+    libgtkmm-3.0-dev git libfftw3-dev \
+    gfortran flex bison wcslib-dev \
+    libboost-numpy-dev liblua5.3-dev
+
+# Build casacore from source
+RUN mkdir /casacore && cd /casacore && git clone -b v3.2.0 https://github.com/casacore/casacore.git src
+RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DBUILD_TESTING="OFF" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
+
+# Get aoflagger
+RUN mkdir /aoflagger && cd /aoflagger && git clone git://git.code.sf.net/p/aoflagger/code src
+RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
+
+# Build latest IDG master from source
+RUN mkdir /idg && cd /idg && git clone https://gitlab.com/astron-idg/idg.git src
+RUN mkdir /idg/build && cd /idg/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
+
+# Build LOFARBeam
+RUN mkdir /lofarbeam && cd /lofarbeam && git clone https://github.com/lofar-astron/LOFARBeam.git src
+RUN mkdir /lofarbeam/build && cd /lofarbeam/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src -DPYTHON_EXECUTABLE=/usr/bin/python3 && make install -j2
 
 ADD . /src
 WORKDIR /src
-
-# Build latest IDG master from source
-RUN mkdir /idg && cd /idg && git clone https://gitlab.com/astron-idg/idg.git  src
-RUN mkdir /idg/build && cd /idg/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
 
 RUN mkdir /build && cd /build && cmake ../src
 RUN cd /build && make -j2

--- a/.travis/Dockerfile_lofar
+++ b/.travis/Dockerfile_lofar
@@ -2,7 +2,7 @@ FROM kernsuite/base:4
 
 RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
     apt-get install -y casacore-data casacore-dev \
-    libboost-python-dev libcfitsio-dev python-dev python-numpy cmake \
+    libboost-python-dev libcfitsio-dev python3-dev python3-numpy cmake \
     build-essential libhdf5-serial-dev libarmadillo-dev lofar-dev \
     libboost-filesystem-dev libboost-system-dev libboost-date-time-dev \
     libboost-signals-dev libboost-program-options-dev libboost-test-dev \

--- a/.travis/Dockerfile_lofar
+++ b/.travis/Dockerfile_lofar
@@ -15,9 +15,9 @@ RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
 RUN mkdir /casacore && cd /casacore && git clone -b v3.2.0 https://github.com/casacore/casacore.git src
 RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DBUILD_TESTING="OFF" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
 
-# Get aoflagger
+# Get aoflagger (don't use -j2 -- travis can't handle it)
 RUN mkdir /aoflagger && cd /aoflagger && git clone git://git.code.sf.net/p/aoflagger/code src
-RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
+RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install
 
 # Build latest IDG master from source
 RUN mkdir /idg && cd /idg && git clone https://gitlab.com/astron-idg/idg.git src

--- a/.travis/Dockerfile_nolofar
+++ b/.travis/Dockerfile_nolofar
@@ -15,9 +15,9 @@ RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
 RUN mkdir /casacore && cd /casacore && git clone -b v3.2.0 https://github.com/casacore/casacore.git src
 RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DBUILD_TESTING="OFF" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
 
-# Get aoflagger
+# Get aoflagger (don't use -j2 -- travis can't handle it)
 RUN mkdir /aoflagger && cd /aoflagger && git clone git://git.code.sf.net/p/aoflagger/code src
-RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
+RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install
 
 ADD . /src
 WORKDIR /src

--- a/.travis/Dockerfile_nolofar
+++ b/.travis/Dockerfile_nolofar
@@ -13,17 +13,17 @@ RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
 
 # Build casacore from source
 RUN mkdir /casacore && cd /casacore && git clone -b v3.2.0 https://github.com/casacore/casacore.git src
-RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j24
+RUN mkdir /casacore/build && cd /casacore/build && cmake -DBUILD_PYTHON3="ON" -DBUILD_PYTHON="OFF" -DUSE_HDF5="ON" -DCXX11="ON" -DBUILD_TESTING="OFF" -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
 
 # Get aoflagger
 RUN mkdir /aoflagger && cd /aoflagger && git clone git://git.code.sf.net/p/aoflagger/code src
-RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j24
+RUN mkdir /aoflagger/build && cd /aoflagger/build && cmake -DCMAKE_INSTALL_PREFIX=/usr ../src && make install -j2
 
 ADD . /src
 WORKDIR /src
 
 RUN mkdir /build && cd /build && cmake ../src
-RUN cd /build && make -j24
-RUN cd /build && make install -j24
+RUN cd /build && make -j2
+RUN cd /build && make install -j2
 RUN DPPP
-RUN cd /build && CTEST_OUTPUT_ON_FAILURE=1 make test -j24
+RUN cd /build && CTEST_OUTPUT_ON_FAILURE=1 make test -j2

--- a/.travis/Dockerfile_nolofar
+++ b/.travis/Dockerfile_nolofar
@@ -2,7 +2,7 @@ FROM kernsuite/base:4
 
 RUN export DEBIAN_FRONTEND=noninteractive && apt-get update && \
     apt-get install -y casacore-data casacore-dev \
-    libboost-python-dev libcfitsio-dev python-dev python-numpy cmake \
+    libboost-python-dev libcfitsio-dev python3-dev python3-numpy cmake \
     build-essential libhdf5-serial-dev libarmadillo-dev \
     libboost-filesystem-dev libboost-system-dev libboost-date-time-dev \
     libboost-signals-dev libboost-program-options-dev libboost-test-dev \

--- a/AOFlaggerStep/test/CMakeLists.txt
+++ b/AOFlaggerStep/test/CMakeLists.txt
@@ -1,5 +1,3 @@
-# $Id: CMakeLists.txt 26355 2013-09-04 11:51:07Z dijkema $
-
 # Do not use lofar_add_test, because libaoflaggerstep should not be linked in.
 add_executable(tAOFlaggerStep tAOFlaggerStep.cc)
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,7 +46,7 @@ find_package(CFITSIO REQUIRED)
 
 #Prevent accidentally finding old BoostConfig.cmake file from casapy
 set(Boost_NO_BOOST_CMAKE ON)
-find_package(Boost COMPONENTS date_time filesystem python program_options system unit_test_framework REQUIRED)
+find_package(Boost COMPONENTS date_time filesystem python3 program_options system unit_test_framework REQUIRED)
 include_directories(${Boost_INCLUDE_DIR})
 
 find_path(LOFAR_STATION_RESPONSE_DIR NAMES StationResponse/Station.h)
@@ -165,8 +165,8 @@ set(DPPP_OBJECT $<TARGET_OBJECTS:DPPP_OBJ>)
 #
 # Dependencies of AOFlagger
 #
+find_package(PythonLibs 3 REQUIRED)
 find_package(PythonInterp REQUIRED)
-find_package(PythonLibs 2.7 REQUIRED)
 find_package(LibXml2 REQUIRED)
 find_package(PNG REQUIRED)
 find_library(FFTW3_LIB fftw3 REQUIRED)
@@ -193,7 +193,7 @@ set(AOFLAGGERSTEP_OBJECT $<TARGET_OBJECTS:AOFlaggerStep_OBJ>)
 set(EXTRA_LIBRARIES ${EXTRA_LIBRARIES}
   ${AOFLAGGER_LIB}
   ${CASACORE_LIBRARIES}
-  ${Boost_DATE_TIME_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY}  ${Boost_PROGRAM_OPTIONS_LIBRARY} ${Boost_PYTHON_LIBRARY}
+  ${Boost_DATE_TIME_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} ${Boost_SYSTEM_LIBRARY}  ${Boost_PROGRAM_OPTIONS_LIBRARY} ${Boost_PYTHON3_LIBRARY}
   ${EXTRA_LIBRARIES}
   ${PYTHON_LIBRARIES}
   ${IDGAPI_LIBRARIES}

--- a/DDECal/test/CMakeLists.txt
+++ b/DDECal/test/CMakeLists.txt
@@ -1,5 +1,3 @@
-# $Id: CMakeLists.txt 38047 2017-07-26 12:49:53Z dijkema $
-
 # Run script to find current directory (needed for getting a tarball from
 # the source directory).
 configure_file(findenv.run_tmpl findenv.run_script)

--- a/DPPP/etc/CMakeLists.txt
+++ b/DPPP/etc/CMakeLists.txt
@@ -1,5 +1,3 @@
-# $Id: CMakeLists.txt 22196 2012-10-03 05:47:41Z diepen $
-
 # Logger configuration
 install(FILES
   DPPP.log_prop

--- a/DPPP_Interpolate/test/CMakeLists.txt
+++ b/DPPP_Interpolate/test/CMakeLists.txt
@@ -1,5 +1,3 @@
-# $Id: CMakeLists.txt 26355 2013-09-04 11:51:07Z dijkema $
-
 include(LofarCTest)
 
 lofar_add_test(tInterpolateStep tInterpolateStep.cc)

--- a/PythonDPPP/CMakeLists.txt
+++ b/PythonDPPP/CMakeLists.txt
@@ -1,5 +1,5 @@
-# If Boost-Python found, build this package. Otherwise give a warning.
-if(${Boost_PYTHON_FOUND})
+# If Boost-Python3 found, build this package. Otherwise give a warning.
+if(${Boost_PYTHON3_FOUND})
 
   include(PythonInstall)
 
@@ -26,12 +26,12 @@ if(${Boost_PYTHON_FOUND})
   python_install(__init__.py DESTINATION lofar/pythondppp)
 
 	find_package(PythonInterp REQUIRED)
-  find_package(PythonLibs 2.6 REQUIRED)
+  find_package(PythonLibs 3 REQUIRED)
   include_directories(${PYTHON_INCLUDE_DIRS})
 
   find_package(Casacore COMPONENTS python)
   
-  find_package(Boost REQUIRED COMPONENTS python)
+  find_package(Boost REQUIRED COMPONENTS python3)
   include_directories(${Boost_INCLUDE_DIR})
 
 #  add_subdirectory(test)

--- a/PythonDPPP/test/CMakeLists.txt
+++ b/PythonDPPP/test/CMakeLists.txt
@@ -1,5 +1,3 @@
-# $Id: CMakeLists.txt 26355 2013-09-04 11:51:07Z dijkema $
-
 include(LofarCTest)
 
 lofar_add_test(tPythonStep)

